// myst-roles@1.7.0/dist/utils.js
import { normalizeLabel } from 'myst-common';
export function classRoleOption(nodeType = 'node') {
    return {
        class: {
            type: String,
            doc: `Annotate the ${nodeType} with a set of space-delimited class names.`,
        },
    };
}
export function labelRoleOption(nodeType = 'node') {
    return {
        label: {
            type: String,
            alias: ['name'],
            doc: `Label the ${nodeType} to be cross-referenced or explicitly linked to.`,
        },
    };
}
/**
 * Adds `class`, `label`
 */
export function commonRoleOptions(nodeType = 'node') {
    return {
        ...classRoleOption(nodeType),
        ...labelRoleOption(nodeType),
    };
}
export function addClassOptions(data, node) {
    if (typeof data.options?.class === 'string') {
        node.class = data.options.class;
    }
    return node;
}
export function addLabelOptions(data, node) {
    const { label, identifier } = normalizeLabel(data.options?.label) || {};
    if (label)
        node.label = label;
    if (identifier)
        node.identifier = identifier;
    return node;
}
export function addCommonRoleOptions(data, node) {
    addClassOptions(data, node);
    addLabelOptions(data, node);
    return node;
}
