// myst-frontmatter@1.9.5/dist/site/types.js
export const PAGE_KNOWN_PARTS = [
    'abstract',
    'summary',
    'keypoints',
    'dedication',
    'epigraph',
    'data_availability',
    'acknowledgments',
];
export const SITE_FRONTMATTER_KEYS = [
    'title',
    'subtitle',
    'short_title',
    'description',
    'thumbnail',
    'thumbnailOptimized',
    'banner',
    'bannerOptimized',
    'tags',
    'authors',
    'reviewers',
    'editors',
    'contributors',
    'venue',
    'github',
    'keywords',
    'affiliations',
    'funding',
    'copyright',
    'options',
    'parts',
    'social',
    ...PAGE_KNOWN_PARTS,
];
export const FRONTMATTER_ALIASES = {
    author: 'authors',
    reviewer: 'reviewers',
    editor: 'editors',
    contributor: 'contributors',
    affiliation: 'affiliations',
    export: 'exports',
    download: 'downloads',
    jupyter: 'thebe',
    part: 'parts',
    ack: 'acknowledgments',
    acknowledgements: 'acknowledgments',
    acknowledgment: 'acknowledgments',
    acknowledgement: 'acknowledgments',
    availability: 'data_availability',
    dataAvailability: 'data_availability',
    'data-availability': 'data_availability',
    quote: 'epigraph',
    plain_language_summary: 'summary',
    'plain-language-summary': 'summary',
    plainLanguageSummary: 'summary',
    lay_summary: 'summary',
    'lay-summary': 'summary',
    keyPoints: 'keypoints',
    key_points: 'keypoints',
    'key-points': 'keypoints',
    image: 'thumbnail',
    identifier: 'identifiers',
    socials: 'social',
};
