// myst-common@1.9.5/dist/ruleids.js
export var RuleId;
(function (RuleId) {
    // Frontmatter rules
    RuleId["validConfigStructure"] = "valid-config-structure";
    RuleId["siteConfigExists"] = "site-config-exists";
    RuleId["projectConfigExists"] = "project-config-exists";
    RuleId["validSiteConfig"] = "valid-site-config";
    RuleId["validProjectConfig"] = "valid-project-config";
    RuleId["configHasNoDeprecatedFields"] = "config-has-no-deprecated-fields";
    RuleId["frontmatterIsYaml"] = "frontmatter-is-yaml";
    RuleId["validPageFrontmatter"] = "valid-page-frontmatter";
    RuleId["validFrontmatterExportList"] = "valid-frontmatter-export-list";
    // Export rules
    RuleId["docxRenders"] = "docx-renders";
    RuleId["jatsRenders"] = "jats-renders";
    RuleId["mdRenders"] = "md-renders";
    RuleId["mecaIncludesJats"] = "meca-includes-jats";
    RuleId["mecaExportsBuilt"] = "meca-exports-built";
    RuleId["mecaFilesCopied"] = "meca-files-copied";
    RuleId["pdfBuildCommandsAvailable"] = "pdf-build-commands-available";
    RuleId["pdfBuildsWithoutErrors"] = "pdf-builds-without-errors";
    RuleId["pdfBuilds"] = "pdf-builds";
    RuleId["texRenders"] = "tex-renders";
    RuleId["exportExtensionCorrect"] = "export-extension-correct";
    RuleId["exportArticleExists"] = "export-article-exists";
    // Parse rules
    RuleId["texParses"] = "tex-parses";
    RuleId["jatsParses"] = "jats-parses";
    RuleId["mystFileLoads"] = "myst-file-loads";
    RuleId["selectedFileIsProcessed"] = "selected-file-is-processed";
    // Directive and role rules
    RuleId["directiveRegistered"] = "directive-registered";
    RuleId["directiveKnown"] = "directive-known";
    RuleId["directiveArgumentCorrect"] = "directive-argument-correct";
    RuleId["directiveOptionsCorrect"] = "directive-options-correct";
    RuleId["directiveBodyCorrect"] = "directive-body-correct";
    RuleId["roleRegistered"] = "role-registered";
    RuleId["roleKnown"] = "role-known";
    RuleId["roleBodyCorrect"] = "role-body-correct";
    // Project structure rules
    RuleId["tocContentsExist"] = "toc-contents-exist";
    RuleId["encounteredLegacyTOC"] = "encountered-legacy-toc";
    RuleId["validTOCStructure"] = "valid-toc-structure";
    RuleId["validTOC"] = "valid-toc";
    // Image rules
    RuleId["imageDownloads"] = "image-downloads";
    RuleId["imageExists"] = "image-exists";
    RuleId["imageFormatConverts"] = "image-format-converts";
    RuleId["imageCopied"] = "image-copied";
    RuleId["imageFormatOptimizes"] = "image-format-optimizes";
    // Math rules
    RuleId["mathLabelLifted"] = "math-label-lifted";
    RuleId["mathEquationEnvRemoved"] = "math-equation-env-removed";
    RuleId["mathEqnarrayReplaced"] = "math-eqnarray-replaced";
    RuleId["mathAlignmentAdjusted"] = "math-alignment-adjusted";
    RuleId["mathRenders"] = "math-renders";
    // Reference rules
    RuleId["referenceTemplateFills"] = "reference-template-fills";
    RuleId["identifierIsUnique"] = "identifier-is-unique";
    RuleId["referenceTargetResolves"] = "reference-target-resolves";
    RuleId["referenceSyntaxValid"] = "reference-syntax-valid";
    RuleId["referenceTargetExplicit"] = "reference-target-explicit";
    RuleId["footnoteReferencesDefinition"] = "footnote-references-definition";
    RuleId["intersphinxReferencesResolve"] = "intersphinx-references-resolve";
    // Link rules
    RuleId["mystLinkValid"] = "myst-link-valid";
    RuleId["sphinxLinkValid"] = "sphinx-link-valid";
    RuleId["rridLinkValid"] = "rrid-link-valid";
    RuleId["rorLinkValid"] = "ror-link-valid";
    RuleId["wikipediaLinkValid"] = "wikipedia-link-valid";
    RuleId["doiLinkValid"] = "doi-link-valid";
    RuleId["linkResolves"] = "link-resolves";
    RuleId["linkTextExists"] = "link-text-exists";
    // Notebook rules
    RuleId["notebookAttachmentsResolve"] = "notebook-attachments-resolve";
    RuleId["notebookOutputCopied"] = "notebook-output-copied";
    // Content rules
    RuleId["mdastSnippetImports"] = "mdast-snippet-imports";
    RuleId["includeContentFilters"] = "include-content-filters";
    RuleId["includeContentLoads"] = "include-content-loads";
    RuleId["gatedNodesJoin"] = "gated-nodes-join";
    RuleId["glossaryUsesDefinitionList"] = "glossary-uses-definition-list";
    RuleId["blockMetadataLoads"] = "block-metadata-loads";
    RuleId["indexEntriesResolve"] = "index-entries-resolve";
    // Citation rules
    RuleId["citationIsUnique"] = "citation-is-unique";
    RuleId["bibFileExists"] = "bib-file-exists";
    RuleId["citationRenders"] = "citation-renders";
    // Code rules
    RuleId["codeMetadataLifted"] = "code-metadata-lifted";
    RuleId["codeMetatagsValid"] = "code-metatags-valid";
    RuleId["codeLangDefined"] = "code-lang-defined";
    RuleId["codeMetadataLoads"] = "code-metadata-loads";
    RuleId["inlineCodeMalformed"] = "inline-code-malformed";
    RuleId["inlineExpressionRenders"] = "inline-expression-renders";
    // Static file rules
    RuleId["staticFileCopied"] = "static-file-copied";
    RuleId["exportFileCopied"] = "export-file-copied";
    RuleId["sourceFileCopied"] = "source-file-copied";
    RuleId["templateFileCopied"] = "template-file-copied";
    RuleId["staticActionFileCopied"] = "static-action-file-copied";
    // Plugins
    RuleId["pluginLoads"] = "plugin-loads";
    // Container rules
    RuleId["containerChildrenValid"] = "container-children-valid";
    // File rules
    RuleId["mystJsonValid"] = "myst-json-valid";
    // Execution
    RuleId["codeCellExecutes"] = "code-cell-executes";
    RuleId["inlineExpressionExecutes"] = "inline-expression-executes";
})(RuleId || (RuleId = {}));
/**
 * Descriptions for each rule ID.
 * TypeScript will enforce that every RuleId has a corresponding description.
 */
export const RULE_ID_DESCRIPTIONS = {
    // Frontmatter rules
    [RuleId.validConfigStructure]: 'Configuration file structure is valid and can be parsed',
    [RuleId.siteConfigExists]: 'Site configuration is found in project',
    [RuleId.projectConfigExists]: 'Project configuration file exists in the directory',
    [RuleId.validSiteConfig]: 'Site configuration passes validation',
    [RuleId.validProjectConfig]: 'Project configuration passes validation',
    [RuleId.configHasNoDeprecatedFields]: 'Configuration uses current field names without deprecated options',
    [RuleId.frontmatterIsYaml]: 'Frontmatter can be parsed as valid YAML',
    [RuleId.validPageFrontmatter]: 'Page frontmatter passes validation',
    [RuleId.validFrontmatterExportList]: 'Export list in frontmatter is valid for the specified format',
    // Export rules
    [RuleId.docxRenders]: 'DOCX document renders without errors',
    [RuleId.jatsRenders]: 'JATS XML renders without errors',
    [RuleId.mdRenders]: 'Markdown output renders without errors',
    [RuleId.mecaIncludesJats]: 'MECA bundle contains required JATS file',
    [RuleId.mecaExportsBuilt]: 'MECA archive builds without errors',
    [RuleId.mecaFilesCopied]: 'MECA files copy to output successfully',
    [RuleId.pdfBuildCommandsAvailable]: 'Required PDF build tools (LaTeX/Typst) are installed',
    [RuleId.pdfBuildsWithoutErrors]: 'PDF compilation completes without LaTeX/Typst errors',
    [RuleId.pdfBuilds]: 'PDF file generates successfully',
    [RuleId.texRenders]: 'LaTeX/Typst document renders without errors',
    [RuleId.exportExtensionCorrect]: 'Export file has the correct extension for its format',
    [RuleId.exportArticleExists]: 'Article file specified in export configuration exists',
    // Parse rules
    [RuleId.texParses]: 'LaTeX content parses without syntax errors',
    [RuleId.jatsParses]: 'JATS XML parses without syntax errors',
    [RuleId.mystFileLoads]: 'MyST markdown file loads and parses successfully',
    [RuleId.selectedFileIsProcessed]: 'File specified for processing is found and processed',
    // Directive and role rules
    [RuleId.directiveRegistered]: 'Directive is registered without name conflicts',
    [RuleId.directiveKnown]: 'Directive name is recognized',
    [RuleId.directiveArgumentCorrect]: 'Directive argument matches specification',
    [RuleId.directiveOptionsCorrect]: 'Directive options are valid and properly formatted',
    [RuleId.directiveBodyCorrect]: 'Directive body content meets requirements',
    [RuleId.roleRegistered]: 'Role is registered without name conflicts',
    [RuleId.roleKnown]: 'Role name is recognized',
    [RuleId.roleBodyCorrect]: 'Role content meets requirements',
    // Project structure rules
    [RuleId.tocContentsExist]: 'Files referenced in table of contents exist',
    [RuleId.encounteredLegacyTOC]: 'Table of contents uses deprecated format',
    [RuleId.validTOCStructure]: 'Table of contents structure passes schema validation',
    [RuleId.validTOC]: 'Table of contents is valid',
    // Image rules
    [RuleId.imageDownloads]: 'Remote image downloads successfully',
    [RuleId.imageExists]: 'Image file exists at specified path',
    [RuleId.imageFormatConverts]: 'Image format converts successfully using available tools',
    [RuleId.imageCopied]: 'Image copies to output directory successfully',
    [RuleId.imageFormatOptimizes]: 'Image format optimizes successfully',
    // Math rules
    [RuleId.mathLabelLifted]: 'Math equation label extracts successfully',
    [RuleId.mathEquationEnvRemoved]: 'Nested equation environment removes successfully',
    [RuleId.mathEqnarrayReplaced]: 'Deprecated eqnarray environment replaces with align',
    [RuleId.mathAlignmentAdjusted]: 'Math alignment adjusts successfully',
    [RuleId.mathRenders]: 'Math expression renders with KaTeX without errors',
    // Reference rules
    [RuleId.referenceTemplateFills]: 'Reference template populates with values successfully',
    [RuleId.identifierIsUnique]: 'Identifier is unique across the project',
    [RuleId.referenceTargetResolves]: 'Cross-reference target is found',
    [RuleId.referenceSyntaxValid]: 'Reference syntax is valid',
    [RuleId.referenceTargetExplicit]: 'Reference target is explicitly specified',
    [RuleId.footnoteReferencesDefinition]: 'Footnote reference links to an existing definition',
    [RuleId.intersphinxReferencesResolve]: 'Intersphinx cross-reference resolves to external documentation',
    // Link rules
    [RuleId.mystLinkValid]: 'MyST-formatted link is valid',
    [RuleId.sphinxLinkValid]: 'Sphinx-style link is valid',
    [RuleId.rridLinkValid]: 'RRID (Research Resource Identifier) link is valid',
    [RuleId.rorLinkValid]: 'ROR (Research Organization Registry) link is valid',
    [RuleId.wikipediaLinkValid]: 'Wikipedia link is valid',
    [RuleId.doiLinkValid]: 'DOI (Digital Object Identifier) link is valid',
    [RuleId.linkResolves]: 'Link URL resolves successfully',
    [RuleId.linkTextExists]: 'Link has non-empty text content',
    // Notebook rules
    [RuleId.notebookAttachmentsResolve]: 'Jupyter notebook attachments resolve and decode successfully',
    [RuleId.notebookOutputCopied]: 'Notebook cell output copies to disk successfully',
    // Content rules
    [RuleId.mdastSnippetImports]: 'MDAST snippet imports successfully',
    [RuleId.includeContentFilters]: 'Include directive filters apply correctly',
    [RuleId.includeContentLoads]: 'Include directive loads target file successfully without circular dependencies',
    [RuleId.gatedNodesJoin]: 'Conditional content nodes merge correctly',
    [RuleId.glossaryUsesDefinitionList]: 'Glossary directive contains a definition list',
    [RuleId.blockMetadataLoads]: 'Block-level metadata loads and parses successfully',
    [RuleId.indexEntriesResolve]: 'Index entries resolve successfully',
    // Citation rules
    [RuleId.citationIsUnique]: 'Citation identifier is unique across bibliography files',
    [RuleId.bibFileExists]: 'Bibliography file specified in configuration exists',
    [RuleId.citationRenders]: 'Citation processes with citation-js successfully',
    // Code rules
    [RuleId.codeMetadataLifted]: 'Code cell metadata extracts without conflicts',
    [RuleId.codeMetatagsValid]: 'Code cell tags are valid',
    [RuleId.codeLangDefined]: 'Code block has a language specified',
    [RuleId.codeMetadataLoads]: 'Code metadata loads successfully',
    [RuleId.inlineCodeMalformed]: 'Inline code has either value or children but not both',
    [RuleId.inlineExpressionRenders]: 'Inline expression evaluates successfully',
    // Static file rules
    [RuleId.staticFileCopied]: 'Static file copies to output successfully',
    [RuleId.exportFileCopied]: 'Export output file copies to public directory successfully',
    [RuleId.sourceFileCopied]: 'Source file copies to output successfully',
    [RuleId.templateFileCopied]: 'Template file copies successfully',
    [RuleId.staticActionFileCopied]: 'Static action file copies successfully',
    // Plugins
    [RuleId.pluginLoads]: 'Plugin loads and executes without errors',
    // Container rules
    [RuleId.containerChildrenValid]: 'Container has valid child elements',
    // File rules
    [RuleId.mystJsonValid]: 'MyST JSON file is valid',
    [RuleId.codeCellExecutes]: 'Code cell executes without errors',
    [RuleId.inlineExpressionExecutes]: 'Inline expression evaluates without errors',
};
