// myst-directives@1.7.0/dist/dropdown.js
import { addCommonDirectiveOptions, commonDirectiveOptions } from './utils.js';
export const dropdownDirective = {
    name: 'dropdown',
    arg: {
        type: 'myst',
    },
    options: {
        ...commonDirectiveOptions('dropdown'),
        // TODO: Add enumeration in future
        open: {
            type: Boolean,
            doc: 'When true, the dropdown starts open.',
        },
        // Legacy options we may want to implement:
        // color
        // icon
        // animate
        // margin
        // name
        // 'class-container'
        // 'class-title'
        // 'class-body'
    },
    body: {
        type: 'myst',
        required: true,
    },
    run(data) {
        const children = [];
        if (data.arg) {
            children.push({ type: 'summary', children: data.arg });
        }
        children.push(...data.body);
        const details = {
            type: 'details',
            open: data.options?.open,
            children,
        };
        addCommonDirectiveOptions(data, details);
        return [details];
    },
};
