// myst-directives@1.7.0/dist/embed.js
import { normalizeLabel } from 'myst-common';
import { commonDirectiveOptions, addCommonDirectiveOptions } from './utils.js';
export const embedDirective = {
    name: 'embed',
    doc: 'The embed directive allows you to duplicate content from another part of your project. This can also be done through the figure directive.',
    arg: {
        type: String,
        doc: 'The label of the node that you are embedding.',
        required: true,
    },
    options: {
        ...commonDirectiveOptions('embed'),
        'remove-input': {
            type: Boolean,
            doc: 'If embedding a Jupyter Notebook cell, remove the input of the cell.',
        },
        'remove-output': {
            type: Boolean,
            doc: 'If embedding a Jupyter Notebook cell, remove the output of the cell.',
        },
    },
    run(data) {
        if (!data.arg)
            return [];
        const argString = data.arg;
        const arg = argString.startsWith('#') ? argString.substring(1) : argString;
        const { label } = normalizeLabel(arg) || {};
        if (!label)
            return [];
        const embed = {
            type: 'embed',
            source: { label },
            'remove-input': data.options?.['remove-input'],
            'remove-output': data.options?.['remove-output'],
        };
        addCommonDirectiveOptions(data, embed);
        return [embed];
    },
};
