// myst-directives@1.7.0/dist/iframe.js
import { addCommonDirectiveOptions, commonDirectiveOptions } from './utils.js';
export const iframeDirective = {
    name: 'iframe',
    arg: {
        type: String,
        doc: 'The URL source (`src`) of the HTML iframe element.',
        required: true,
    },
    options: {
        ...commonDirectiveOptions('iframe'),
        width: {
            type: String,
            doc: 'The iframe width, in CSS units, for example `50%` or `300px`.',
        },
        align: {
            type: String,
            doc: 'The alignment of the iframe in the page. Choose one of `left`, `center` or `right`',
        },
        title: {
            type: String,
            doc: 'The title attribute for the iframe element, describing its content for accessibility.',
        },
        placeholder: {
            type: String,
            doc: 'A placeholder image for the iframe in static exports.',
        },
    },
    body: { type: 'myst', doc: 'If provided, this will be the iframe caption.' },
    run(data) {
        const iframe = {
            type: 'iframe',
            src: data.arg,
            width: data.options?.width,
            align: data.options?.align,
            title: data.options?.title,
        };
        if (data.options?.placeholder) {
            iframe.children = [
                {
                    type: 'image',
                    placeholder: true,
                    url: data.options.placeholder,
                    alt: data.options?.alt,
                    width: data.options?.width,
                    height: data.options?.height,
                    align: data.options?.align,
                },
            ];
        }
        if (!data.body) {
            addCommonDirectiveOptions(data, iframe);
            return [iframe];
        }
        const container = {
            type: 'container',
            kind: 'figure',
            children: [iframe, { type: 'caption', children: data.body }],
        };
        addCommonDirectiveOptions(data, container);
        return [container];
    },
};
