// myst-directives@1.7.0/dist/indices.js
import { createIndexEntries, fileError, parseIndexLine } from 'myst-common';
function warnOnOptionSyntax(option, value, vfile, node) {
    fileError(vfile, `Index entry definitions should not use :option: syntax`, {
        node,
        note: `Replace ":${option}: ${value}" with "${option}: ${value}"`,
    });
}
export const indexDirective = {
    name: 'index',
    arg: {
        type: String,
    },
    options: {
        single: {
            type: String,
        },
        pair: {
            type: String,
        },
        triple: {
            type: String,
        },
        see: {
            type: String,
        },
        seealso: {
            type: String,
            alias: ['seeAlso', 'see-also'],
        },
        label: {
            type: String,
            alias: ['name'],
        },
    },
    body: {
        type: String,
    },
    run(data, vfile) {
        const values = { single: [], pair: [], triple: [], see: [], seealso: [] };
        if (data.arg)
            parseIndexLine(data.arg, values, vfile, data.node);
        if (data.options?.single) {
            warnOnOptionSyntax('single', data.options.single, vfile, data.node);
            values.single.push(data.options.single);
        }
        if (data.options?.pair) {
            warnOnOptionSyntax('pair', data.options.pair, vfile, data.node);
            values.pair.push(data.options.pair);
        }
        if (data.options?.triple) {
            warnOnOptionSyntax('triple', data.options.triple, vfile, data.node);
            values.triple.push(data.options.triple);
        }
        if (data.options?.see) {
            warnOnOptionSyntax('see', data.options.see, vfile, data.node);
            values.see.push(data.options.see);
        }
        if (data.options?.seealso) {
            warnOnOptionSyntax('seealso', data.options.seealso, vfile, data.node);
            values.seealso.push(data.options.seealso);
        }
        if (data.body) {
            data.body.split('\n').forEach((line) => {
                parseIndexLine(line, values, vfile, data.node);
            });
        }
        const entries = createIndexEntries(values, vfile, data.node);
        const output = [
            {
                type: 'mystTarget',
                label: data.options?.label,
                indexEntries: entries,
            },
        ];
        return output;
    },
};
export const genIndexDirective = {
    name: 'show-index',
    alias: ['genindex'],
    arg: {
        type: 'myst',
        doc: 'Heading to be included in index block',
    },
    run(data) {
        const children = [];
        if (data.arg) {
            const parsedArg = data.arg;
            if (parsedArg[0]?.type === 'heading') {
                children.push(...parsedArg);
            }
            else {
                children.push({
                    type: 'heading',
                    depth: 2,
                    enumerated: false,
                    children: parsedArg,
                });
            }
        }
        return [{ type: 'genindex', children }];
    },
};
