// myst-directives@1.7.0/dist/mystdemo.js
import yaml from 'js-yaml';
import { addCommonDirectiveOptions, commonDirectiveOptions } from './utils.js';
export const mystdemoDirective = {
    name: 'myst',
    doc: 'Demonstrate some MyST code in an editable code block that displays its output/result interactively. Limited to built-in functionality, as parsing and rendering is done in the browser without access to plugin code or project configuration.',
    options: {
        ...commonDirectiveOptions('myst'),
        numbering: {
            type: String,
        },
    },
    body: {
        type: String,
        required: true,
    },
    run(data) {
        let numbering;
        if (data.options?.numbering) {
            try {
                numbering = yaml.load(data.options.numbering);
            }
            catch (err) {
                //pass
            }
        }
        const myst = {
            type: 'myst',
            numbering,
            value: data.body,
        };
        addCommonDirectiveOptions(data, myst);
        return [myst];
    },
};
