// myst-directives@1.7.0/dist/toc.js
import { fileError } from 'myst-common';
import { addCommonDirectiveOptions, commonDirectiveOptions } from './utils.js';
const CONTEXTS = ['project', 'children', 'page', 'section'];
export const tocDirective = {
    name: 'toc',
    doc: 'Inserts table of contents in the page. This may be for every page of the project, children of the current page, headings on the current page, or headings in the current section.',
    alias: ['tableofcontents', 'table-of-contents', 'toctree', 'contents'],
    arg: {
        type: 'myst',
        doc: 'Heading to be included with table of contents',
    },
    options: {
        context: {
            type: String,
            doc: 'Table of Contents context; one of project, children, page, or section',
            alias: ['kind'],
        },
        depth: {
            type: Number,
            doc: 'Number of levels to include in Table of Contents; by default, all levels will be included',
            alias: ['maxdepth'],
        },
        ...commonDirectiveOptions('toc'),
    },
    run(data, vfile) {
        let context = data.options?.context
            ? data.options.context
            : data.name === 'contents'
                ? 'section'
                : 'project';
        if (!CONTEXTS.includes(context)) {
            fileError(vfile, `Unknown context for ${data.name} directive: ${context}`);
            context = 'project';
        }
        let depth = data.options?.depth;
        if (depth != null && depth < 1) {
            fileError(vfile, `Table of Contents 'depth' must be a number greater than 0`);
            depth = undefined;
        }
        const children = [];
        if (data.arg) {
            const parsedArg = data.arg;
            if (parsedArg[0]?.type === 'heading') {
                children.push(...parsedArg);
            }
            else {
                children.push({
                    type: 'heading',
                    depth: 2,
                    enumerated: false,
                    children: parsedArg,
                });
            }
        }
        const toc = { type: 'toc', kind: context, depth, children };
        addCommonDirectiveOptions(data, toc);
        return [toc];
    },
};
